src/ContentLoader.js

```diff
@@ -0,0 +1,46 @@
+import React, {Component, PropTypes, createElement} from 'react';
+
+import Loader from './components/Page/Loader';
+import PageRenderer from './components/Page/PageRenderer';
+import Page from './components/Page/Page';
+
+class ContentLoader extends Component {
+  constructor() {
+    super();
+    this.state = {content: null};
+  }
+
+  componentWillMount() {
+    this.fetchPageContent();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.urlOrComponentPromise !== this.props.urlOrComponentPromise) {
+      this.fetchPageContent();
+    }
+  }
+
+  fetchPageContent() {
+    const {urlOrComponentPromise} = this.props;
+
+    const contentPromise = typeof urlOrComponentPromise === 'string'
+      ? fetch(urlOrComponentPromise, {credentials: 'same-origin'}).then(res => res.text()).then(text => <Page>{text}</Page>)
+      : urlOrComponentPromise().then(c => createElement(c));
+
+    contentPromise
+      .then((content) => { this.setState({content}); })
+      .catch((content) => { this.setState({content}); });
```

Comment:
1. This should be the two-argument form of `.then(onSuccess, onError)` instead of `.catch` because [we don't want to catch React rendering errors](https://twitter.com/dan_abramov/status/776891811222675457).
2. Probably we also want to [reject response statuses outside of the `2xx` range](https://github.com/github/fetch#handling-http-error-statuses) (_before_ calling `.then(res => res.text())`)
3. When an error happens, this now just displays the raw error, right? That should probably be a nice message :)